docs/PROTOCOL_v21.md

```diff
@@ -0,0 +1,375 @@
+
+### Samples Observer
+
+Processes samples of upkeeps to be checked, using the latest block number as a trigger. It does the following procedures:
+
+- Pre-processes to filter upkeep present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into **metadata store**
+with conditional Upkeep instructions
+- Errors and ineligible results are ignored
+
+### Conditional Observer
+
+Processes coordinated block + upkeep payload coming from plugin and does the following:
+
+- Pre-processes to filter upkeep present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into **result store**
+- Errors and ineligible results are ignored
+
+### Log Observer
+
+Processes latest logs for active upkeeps. It does the following procedures:
+
+- Called every second by Log Trigger Ticker with latest logs as full payloads.
+- Filter out logs with block number that is older than threshold
+    - These are expected to be picked up by the log recoverer 
+- Pre-processes to filter the logs already present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into result store
+- If runner gave an error, put back into log retry ticker, to be performed after a fixed timeout (can extend to exponential backoff)
+
+<aside>
+💡 Note: Duplicate logs can enter this flow, this component should be able to handle that. This component can maintain in-memory state for processed logs and can also use that state for exponential backoff
+</aside>
+
+### Retry Observer
+
+**TODO**
+
+### Recovery Observer
+
+This flow gets as input an upkeepID and a logIdentifier. It is a backup flow which verifies that a log was fully processed. Logs can be put here from the node itself (through Log Recoverer) or from other nodes (through Plugin)
+
+- Input is an upkeepID, logTxHash, logIndex (optional checkBlockNum, checkBlockHash when passed from plugin)
+- Checks whether log should be processed
+    - Whether its part of the filters for that upkeep and it is still present on chain (DB call to log poller)
+    - Whether it is older from the latest block within a threshold (To prevent malicious node requesting recovery on a new log)
+- Checks whether the log has been already processed
+    - If node locally checked (logTxHash, logIndex) and got a non eligible result then skip (DB call)
+    - If chain has a perform for the log then skip (DB call which looks at upkeepPerformed logs)
+- If checkBlockNum not present: puts into an **in-memory recovery queue** which is then read in the plugin and put back in the flow with a coordinated checkBlock
+- If checkBlockNum present
+    - get payload from log provider for (upkeepID, logTxHash, logIndex)
+    - call runner with payload
+    - upon result, if eligible add into result store
+    - non-eligible are written to DB
+    - If runner gave an error then it is ignored (it might get picked up again by log recoverer, but we do not keep retrying here)
```

Comment:
The plugin does continue retrying by putting the retryable result back into the recovery proposal flow.

Reply:
Ended the flow here. The log will be picked up again by log recoverer